toolkit/tools/pkg/imagecustomizerlib/customizeboot.go

```diff
@@ -107,143 +93,228 @@ func replaceToken(inputGrubCfgContent string, oldToken string, newToken string)
+		}
+
+		if hasVarExpansion {
+			// The arg string value isn't known because it contains a variable expansion.
+			value = ""
+		}
+
+		arg := grubConfigLinuxArg{
+			Token:                argToken,
+			Name:                 name,
+			Value:                value,
+			ValueHasVarExpansion: hasVarExpansion,
+		}
+		args = append(args, arg)
 	}
 
-	if start == -1 {
-		return "", "", fmt.Errorf("failed to find kernel arg value for (%s)", name)
+	if insertAtToken == nil {
+		return nil, grub.Token{}, fmt.Errorf("failed to find $kernelopts in linux command line")
 	}
 
-	outputGrubCfgContent = fmt.Sprintf("%s%s=%s%s", inputGrubCfgContent[:start], name, value, inputGrubCfgContent[end:])
+	return args, *insertAtToken, nil
+}
 
-	return outputGrubCfgContent, oldValue, nil
+// Filters a list of kernel command-line args to only those that match the provided list of names.
+func findMatchingCommandLineArgs(args []grubConfigLinuxArg, names []string) []grubConfigLinuxArg {
+	matching := []grubConfigLinuxArg(nil)
+
+argsLoop:
+	for _, arg := range args {
+		for _, name := range names {
+			if arg.Name == name {
+				matching = append(matching, arg)
+				continue argsLoop
+			}
+		}
+	}
+
+	return matching
 }
 
-// Tries to remove the specified kernel CLI arg from the grub file. Does not fail if the arg is not present.
+// Tries to find the specified kernel CLI arg. Does not fail if the arg is not found.
 //
 // Returns:
-// - outputGrubCfgContent: The resulting grub.cfg file.
-// - index: The index in the grub.cfg where the arg was removed from. Or -1 if not found.
-func tryRemoveKernelCommandLineArgumentValue(inputGrubCfgContent string, name string) (string, int, error) {
-	// Search for the kernel CLI arg.
-	start, end, _, err := findKernelCommandLineArgumentArg(inputGrubCfgContent, name)
+// - value: The value of the arg. Or "" if not found.
+func findKernelCommandLineArgValue(args []grubConfigLinuxArg, name string) (string, error) {
+	foundArgs := findMatchingCommandLineArgs(args, []string{name})
+	if len(foundArgs) <= 0 {
+		return "", nil
+	}
+
+	lastArg := foundArgs[len(foundArgs)-1]
+	if lastArg.ValueHasVarExpansion {
+		return "", fmt.Errorf("kernel arg (%s) has variable expansion in value", name)
+	}
+
+	return lastArg.Value, nil
+}
+
+// Finds an existing kernel command-line arg and replaces its value.
+func replaceKernelCommandLineArgumentValue(inputGrubCfgContent string, name string, value string,
+) (outputGrubCfgContent string, oldValue string, err error) {
+	newArg := fmt.Sprintf("%s=%s", name, value)
+	quotedNewArg := grub.QuoteString(newArg)
+
+	args, _, err := getLinuxCommandLineArgs(inputGrubCfgContent)
 	if err != nil {
-		return "", 0, err
+		return "", "", err
 	}
 
-	if start == -1 {
-		// Arg not found.
-		return inputGrubCfgContent, -1, nil
+	foundArgs := findMatchingCommandLineArgs(args, []string{name})
+	if len(foundArgs) < 1 {
+		return "", "", fmt.Errorf("failed to find kernel arg (%s)", name)
 	}
+	if len(foundArgs) > 1 {
```

Comment:
Also, here, can we just eliminate all those duplicate instances and replacement them with our new single instance?

Reply:
The `updateKernelCommandLineArguments` function also already handles removing duplicate args.